accio-sqlrewrite/src/main/java/io/accio/sqlrewrite/AccioSqlRewrite.java

```diff
@@ -54,20 +57,55 @@ public class AccioSqlRewrite
 
     private AccioSqlRewrite() {}
 
+    private LinkedHashMap<String, Set<String>> getRequiredFields(AccioDataLineage dataLineage, Analysis analysis)
+    {
+        List<QualifiedName> collectedColumns = analysis.getCollectedColumns().asMap().entrySet().stream()
+                .map(e ->
+                        e.getValue().stream()
+                                .map(columnName -> QualifiedName.of(e.getKey().getSchemaTableName().getTableName(), columnName))
+                                .collect(toImmutableList()))
+                .flatMap(List::stream)
+                .collect(toImmutableList());
+        return dataLineage.getRequiredFields(collectedColumns);
+    }
+
     @Override
     public Statement apply(Statement root, SessionContext sessionContext, Analysis analysis, AccioMDL accioMDL)
     {
-        Set<QueryDescriptor> modelDescriptors = analysis.getModels().stream().map(model -> RelationInfo.get(model, accioMDL)).collect(toSet());
-        Set<QueryDescriptor> metricDescriptors = analysis.getMetrics().stream().map(metric -> RelationInfo.get(metric, accioMDL)).collect(toSet());
-        Set<QueryDescriptor> cumulativeMetricDescriptors = analysis.getCumulativeMetrics().stream().map(metric -> CumulativeMetricInfo.get(metric, accioMDL)).collect(toSet());
-        Set<QueryDescriptor> viewDescriptors = analysis.getViews().stream().map(view -> ViewInfo.get(view, accioMDL, sessionContext)).collect(toSet());
-        Set<QueryDescriptor> allDescriptors = ImmutableSet.<QueryDescriptor>builder()
-                .addAll(modelDescriptors)
-                .addAll(metricDescriptors)
-                .addAll(viewDescriptors)
-                .addAll(cumulativeMetricDescriptors)
-                .build();
-        return apply(root, sessionContext, analysis, accioMDL, allDescriptors);
+        Set<QueryDescriptor> allDescriptors;
+        // TODO: Currently DynamicCalculatedField is a experimental feature, and buggy. After all issues are solved,
+        //  we should always enable this setting.
+        if (sessionContext.isEnableDynamicCalculatedField()) {
+            // TODO: make AccioDataLineage static instead of create a new one everytime as it will change only when accio mdl changed.
+            AccioDataLineage dataLineage = AccioDataLineage.analyze(accioMDL);
```

Comment:
How about analyzing data lineage when construct `AccioMDL`? Make `AccioDataLineage` be a member of `AccioMDL`. We don't need to analyze this everytime.

Reply:
as TODO mentioned, I'll to this in other PRs.

> Make AccioDataLineage be a member of AccioMDL. We don't need to analyze this everytime.

Currently AccioDataLineage depends on jgrapht and some code in accio-sqlrewrite. If we are going to do this. We will would require some works to do. So as TODO mentioned, I'll open another PR to solve this.